### PR TITLE
Playground state

### PR DESCRIPTION
**Pull request by feloy**
With this PR, in a Playground page, the query and the response are decorrelated.

- the frontend sends a query and gets a queryId in return
- the backend waits for the query to terminate, and updates the queries state with the result of the query
- the frontend listens for the query state (through the playgroundQuery store), and updates the page with the result.
- if the user leaves the Playground page and goes back later to the page, the latest query is displayed (either it is terminated or waiting for a response)

Other related changes:
- the frontend can subscribe to messages from the RpcBrowser (which receives all messages from backend)
- the container for the playground is not restarted if already started
- a new port is used for every new playground's container, it should be possible to have several playground's containers running at the same time (to be tested)

TODO
- [x] tests

### CODE REVIEW

**Comment by slemeur** [comment]
👍

File: packages/backend/src/studio-api-impl.ts [hunk 1]
```diff
@@ -111,11 +111,15 @@ export class StudioApiImpl implements StudioAPI {
...
     const localModelInfo = this.applicationManager.getLocalModels().filter(m => m.id === modelId);
     if (localModelInfo.length !== 1) {
       throw new Error('model not found');
     }
     return this.playgroundManager.askPlayground(localModelInfo[0], prompt);
   }
+
+  async askPlaygroundQueries(): Promise<void> {
+    this.playgroundManager.sendState();
+  }
```
Comment (axel7083) [inline packages/backend/src/studio-api-impl.ts:124]:
```suggestion
  async getPlaygroundStates(): Promise<QueryState[]> {
    return this.playgroundManager.getState();
  }
```

File: packages/backend/src/playground.ts [hunk 2]
```diff
@@ -69,42 +97,72 @@ export class PlayGroundManager {
...
+  }
+
+  getNextQueryId() {
+    return ++this.queryIdCounter;
+  }
+  
+  sendState() {
+    this.webview.postMessage({ 
+      id: MSG_NEW_PLAYGROUND_QUERIES_STATE, 
+      body: Array.from(this.queries.values()),
```
Comment (axel7083) [inline packages/backend/src/playground.ts:165]:
```suggestion
      body: this.getState(),
```

File: packages/backend/src/playground.ts [hunk 3]
```diff
@@ -69,42 +97,72 @@ export class PlayGroundManager {
...
+
+    this.queries.set(query.id, query);
+    this.sendState();
+    return query.id;
+  }
+
+  getNextQueryId() {
+    return ++this.queryIdCounter;
+  }
+
```
Comment (axel7083) [inline packages/backend/src/playground.ts:161]:
```suggestion
  getState(): QueryState[] {
    return Array.from(this.queries.values());
  }
```

File: packages/shared/StudioAPI.ts [hunk 4]
```diff
@@ -23,12 +23,17 @@ export abstract class StudioAPI {
...
   /**
    * Get task by label
    * @param label
    */
   abstract getTasksByLabel(label: string): Promise<Task[]>;
+
+  /**
+   * Ask to send a message MSG_NEW_PLAYGROUND_QUERIES_STATE with the current Playground queries
+   */
+  abstract askPlaygroundQueries(): Promise<void>;
```
Comment (axel7083) [inline packages/shared/StudioAPI.ts:37]:
```suggestion
  abstract getPlaygroundStates(): Promise<QueryState[]>;
```

File: packages/frontend/src/stores/playground-queries.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+import type { QueryState } from '@shared/models/IPlaygroundQueryState';
+import { rpcBrowser } from '../utils/client';
+import { MSG_NEW_PLAYGROUND_QUERIES_STATE } from '@shared/Messages';
+import { studioClient } from '/@/utils/client';
+
+export const playgroundQueries: Readable<QueryState[]> = readable<QueryState[]>([], (set) => {  
+  const sub = rpcBrowser.subscribe(MSG_NEW_PLAYGROUND_QUERIES_STATE, (msg) => {
+    set(msg);
+  });
+  studioClient.askPlaygroundQueries();
```
Comment (axel7083) [inline packages/frontend/src/stores/playground-queries.ts:12]:
```suggestion
  // Initialize the store manually 
  studioClient.getPlaygroundStates().then((state) => {
    set(state);
  });
```

File: packages/shared/MessageProxy.ts [hunk 6]
```diff
@@ -153,7 +166,23 @@ export class RpcBrowser {
...
     // Create a Promise
     return new Promise((resolve, reject) => {
-      this.promises.set(requestId, {resolve, reject});
+      this.promises.set(requestId, { resolve, reject });
     })
   }
+
+  // TODO(feloy) need to subscribe several times?
+  subscribe(msgId: string, f: (msg: any) => void): Subscriber {
+    console.log('add subscriber');
```
Comment (axel7083) [inline packages/shared/MessageProxy.ts:175]:
```suggestion
```

File: packages/shared/MessageProxy.ts [hunk 7]
```diff
@@ -153,7 +166,23 @@ export class RpcBrowser {
...
+    this.subscribers.set(msgId, f);
+    return {
+      unsubscribe: () => {
+        console.log('delete subscriber');
```
Comment (axel7083) [inline packages/shared/MessageProxy.ts:179]:
```suggestion
```

**Review by axel7083 — APPROVED** [review]
Just an idea, instead of having a method firing the first event to setup the store, we could make it return the state, and initiate the store with it, it would look less _magic_ to read

**Review by lstocchi — COMMENTED** [review]
Idk if it would be part of this PR or a separate one but the playground does not work for me. I don't have the local-ai image and i don't know what else i'm missing (still have to check the code) but as a user i clicked `send request` and nothing happens. If i look at the console i see 
image

**Comment by feloy** [comment]
> Idk if it would be part of this PR or a separate one but the playground does not work for me. I don't have the local-ai image and i don't know what else i'm missing (still have to check the code) but as a user i clicked `send request` and nothing happens. If i look at the console i see image

I think it is due to the fact that you don't have the local-ai image.
I have created this issue to follow the problem: https://github.com/projectatomic/studio-extension/issues/87

**Review by lstocchi — APPROVED** [review]
After having downloaded local-ai, it works as expected. GJ 🚀

File: packages/backend/src/playground.ts [hunk 8]
```diff
@@ -15,13 +24,33 @@ function findFirstProvider(): ProviderContainerConnection | undefined {
...
+    this.queries = new Map<number, QueryState>();
+  }
+
   async selectImage(connection: ProviderContainerConnection, image: string): Promise<ImageInfo | undefined> {
     const images = (await containerEngine.listImages()).filter(im => im.RepoTags?.some(tag => tag === image));
     return images.length > 0 ? images[0] : undefined;
   }
 
   async startPlayground(modelId: string, modelPath: string): Promise<string> {
+    // TODO(feloy) remove previous query from state?
```
Comment (axel7083) [inline packages/backend/src/playground.ts:49]: todo ?

Comment (feloy) [reply packages/backend/src/playground.ts:49]: not as part of this PR

File: packages/shared/src/MessageProxy.ts [hunk 9]
```diff
@@ -183,4 +196,24 @@ export class RpcBrowser {
       this.promises.set(requestId, { resolve, reject });
     });
   }
+
+  // TODO(feloy) need to subscribe several times?
```
[the remark this answers is not in the document]
Comment (feloy) [reply packages/shared/src/MessageProxy.ts:200]: not as part of this PR

**Pull request merged**